File: Book.m
#import "Book.h"
#import "Ævintýri-swift.h"

@implementation Book

-(id)initWithBookFolderURL: (NSURL *) bookFolderUrl {

    if (self=[super init]) {
        
        self.bookUrl = bookFolderUrl;
        self.title = bookFolderUrl.lastPathComponent;
        
        NSArray *folderContents = [[NSFileManager defaultManager] contentsOfDirectoryAtURL:bookFolderUrl includingPropertiesForKeys:nil options:0 error:nil];
    
        NSURL *tmpurl = [bookFolderUrl URLByAppendingPathComponent:@"icon.jpg"];
        if ([tmpurl checkResourceIsReachableAndReturnError:nil]) {
            self.icon = tmpurl;
        }
        
        NSPredicate *fltr = [NSPredicate predicateWithFormat:@"lastPathComponent contains 'audio'"];
        NSArray *pageAudioURLs = [folderContents filteredArrayUsingPredicate:fltr];
        pageAudioURLs = [pageAudioURLs sortedArrayUsingSelector:@selector(lastPathComponent)];
        pageAudioURLs = [pageAudioURLs sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
            return [[obj1 lastPathComponent] compare:[obj2 lastPathComponent] options:NSNumericSearch];
        }];
        
        //all the files that contain the word image
        fltr = [NSPredicate predicateWithFormat:@"lastPathComponent contains 'image'"];
        NSArray *pageImageURLs = [folderContents filteredArrayUsingPredicate:fltr];
        pageImageURLs = [pageImageURLs sortedArrayUsingSelector:@selector(lastPathComponent)];
        pageImageURLs = [[pageImageURLs reverseObjectEnumerator] allObjects];
        pageImageURLs = [pageImageURLs sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
            return [[obj1 lastPathComponent] compare:[obj2 lastPathComponent] options:NSNumericSearch];
        }];

        //all the files that contain the word text
        fltr = [NSPredicate predicateWithFormat:@"lastPathComponent contains 'text'"];
        NSArray *pageTextURLs = [folderContents filteredArrayUsingPredicate:fltr];
        pageTextURLs = [pageTextURLs sortedArrayUsingSelector:@selector(lastPathComponent)];
                         
        pageTextURLs = [pageTextURLs sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
            return [[obj1 lastPathComponent] compare:[obj2 lastPathComponent] options:NSNumericSearch];
        }];
        

        int numberOfPages =(int) MAX( MAX(pageImageURLs.count, pageAudioURLs.count), pageTextURLs.count);
        
        for (int iterator=0; iterator<numberOfPages; iterator++) {
            NSURL *imageURL = iterator < pageImageURLs.count ? [pageImageURLs objectAtIndex:iterator] : nil;
            NSURL *audioURL = iterator < pageAudioURLs.count ? [pageAudioURLs objectAtIndex:iterator] : nil;
            NSURL *textURL = iterator < pageTextURLs.count ? [pageTextURLs objectAtIndex:iterator] : nil;
            
            Page *tmpPage = [[Page alloc] initWithImage:imageURL audio:audioURL text:textURL book:self];
            [self.pages addObject:tmpPage];
        }
        
        tmpurl = [bookFolderUrl URLByAppendingPathComponent:@"cover.jpg"];
        if ([tmpurl checkResourceIsReachableAndReturnError:nil]) {
            self.coverURL = [pageImageURLs objectAtIndex:0];
        }else{
            self.coverURL = tmpurl;
        }
        
        return self;
    }
    else{
        return nil;
    }
}



@synthesize pages = _pages;

-(NSMutableArray *) pages{
    if (!_pages) {
        _pages = [[NSMutableArray alloc] init];
    }
    return _pages;
}

@end

File: Book.swift
import Foundation

class Book {
    var pages: [Page] = []
    var icon: URL?
    let sourceFolder: URL
    let title: String
    
    init?(sourceFolder: URL) {
        var isDirectory: ObjCBool = false
        if FileManager.default.fileExists(atPath: sourceFolder.path,
                                          isDirectory: &isDirectory) && isDirectory.boolValue {
            self.sourceFolder = sourceFolder
            self.title = sourceFolder.lastPathComponent
            self.icon = self.getFilesWithPredicate(predicate: "icon").first
            self.pages = self.createPages()
        } else {
            return nil
        }
    }
    
    func createPages() -> [Page] {
        let images: [URL] = getFilesWithPredicate(predicate: "image")
        let audios: [URL] = getFilesWithPredicate(predicate: "audio")
        let texts: [URL] = getFilesWithPredicate(predicate: "text")
        let numberOfPages = ([images.count, audios.count, texts.count].max() != nil) ?
            [images.count, audios.count, texts.count].max()! : 0
        
        var tempPages: [Page] = []
        for _ in 0 ..< numberOfPages {
            tempPages.append(Page(image: nil, audio: nil, text: nil, book: self))
        }
        for (index, image) in images.enumerated() {
            tempPages[index].imageURL = image
        }
        for (index, audio) in audios.enumerated() {
            tempPages[index].audioURL = audio
        }
        for (index, text) in texts.enumerated() {
            tempPages[index].textURL = text
        }
        
        return tempPages
    }
    
    func getFilesWithPredicate(predicate: String) -> [URL] {
        var result: [URL] = []
        do {
            let fileManager = FileManager.default
            let folderContents = try fileManager.contentsOfDirectory(at: sourceFolder, includingPropertiesForKeys: nil,
                                                                     options: .skipsHiddenFiles)
            
            result = (folderContents.compactMap { (url) -> URL? in
                url.lastPathComponent.hasPrefix(predicate) ? url : nil
            }).sorted { (first, second) -> Bool in
                first.lastPathComponent < second.lastPathComponent
            }
        } catch {
            print("Unexpected error: \(error).")
        }
        return result
    }
    
    func pageBeforePage(page: Page) -> Page? {
        var previousPage: Page?
        let indexOfPage = self.pages.firstIndex { (inPage) -> Bool in
            inPage === page
        }
        
        if let indexOfPage = indexOfPage, indexOfPage > 0 {
            previousPage = self.pages[indexOfPage - 1]
        }
        return previousPage
    }
    
    func pageAfterPage(page: Page) -> Page? {
        var nextPage: Page?
        let indexOfPage = self.pages.firstIndex { (inPage) -> Bool in
            inPage === page
        }
        
        if let indexOfPage = indexOfPage, indexOfPage < self.pages.endIndex - 1 {
            nextPage = self.pages[indexOfPage + 1]
        }
        return nextPage
    }
}
